Sources/FoundationEssentials/Data/ContiguousBytes.swift

```diff
@@ -13,16 +13,57 @@
 //===--- ContiguousBytes --------------------------------------------------===//
 
 /// Indicates that the conforming type is a contiguous collection of raw bytes
-/// whose underlying storage is directly accessible by withUnsafeBytes.
+/// whose underlying storage is directly accessible by withBytes.
 @available(macOS 10.10, iOS 8.0, watchOS 2.0, tvOS 9.0, *)
-public protocol ContiguousBytes {
+public protocol ContiguousBytes: ~Escapable, ~Copyable {
+#if !hasFeature(Embedded)
     /// Calls the given closure with the contents of underlying storage.
     ///
     /// - note: Calling `withUnsafeBytes` multiple times does not guarantee that
     ///         the same buffer pointer will be passed in every time.
     /// - warning: The buffer argument to the body should not be stored or used
     ///            outside of the lifetime of the call to the closure.
     func withUnsafeBytes<R>(_ body: (UnsafeRawBufferPointer) throws -> R) rethrows -> R
+#else
+    /// Calls the given closure with the contents of underlying storage.
+    ///
+    /// - note: Calling `withUnsafeBytes` multiple times does not guarantee that
+    ///         the same buffer pointer will be passed in every time.
+    /// - warning: The buffer argument to the body should not be stored or used
+    ///            outside of the lifetime of the call to the closure.
+    func withUnsafeBytes<R, E>(_ body: (UnsafeRawBufferPointer) throws(E) -> R) throws(E) -> R
+#endif
+
+    /// Calls the given closure with the contents of underlying storage.
+    ///
+    /// - note: Calling `withBytes` multiple times does not guarantee that
+    ///         the same span will be passed in every time.
+    @available(FoundationPreview 6.3, *)
+    func withBytes<R, E>(_ body: (RawSpan) throws(E) -> R) throws(E) -> R
+}
+
+extension ContiguousBytes where Self: ~Escapable, Self: ~Copyable {
+    /// Calls the given closure with the contents of underlying storage.
+    ///
+    /// - note: Calling `withBytes` multiple times does not guarantee that
+    ///         the same span will be passed in every time.
+    @_alwaysEmitIntoClient
+    public func withBytes<R, E>(_ body: (RawSpan) throws(E) -> R) throws(E) -> R {
+#if !hasFeature(Embedded)
+        do {
+            return try withUnsafeBytes { (buffer) in
+                try body(buffer.bytes)
+            }
+        } catch let error {
+            // Note: withUnsafeBytes is rethrowing, so we have an "any Error" here that needs casting.
+            throw error as! E
+        }
+#else
+        return try withUnsafeBytes { (buffer) throws(E) in
+            try body(buffer.bytes)
+        }
+#endif
+    }
 }
 
 //===--- Collection Conformances ------------------------------------------===//
```

Comment:
Should the [array types][] implement `withBytes(_:)` instead of using the default implementation?

[array types]: <https://github.com/swiftlang/swift-foundation/blob/swift-6.2.1-RELEASE/Sources/FoundationEssentials/Data/ContiguousBytes.swift#L28-L40>